jax/scipy/optimize/bfgs_minimize.py

```diff
@@ -0,0 +1,127 @@
+"""The Broyden-Fletcher-Goldfarb-Shanno minimization algorithm.
+https://pages.mtu.edu/~struther/Courses/OLD/Sp2013/5630/Jorge_Nocedal_Numerical_optimization_267490.pdf
+"""
+
+import jax
+import jax.numpy as np
+from jax.lax import while_loop
+from .line_search import line_search
+import collections
+
+BFGSResults = collections.namedtuple(
+    'BFGSResults', [
+        'converged',  # bool, True if minimization converges
+        'failed',  # bool, True if line search fails
+        'k',  # The number of iterations of the BFGS update.
+        'nfev',  # The total number of objective evaluations performed.
+        'ngev',  # total number of jacobian evaluations
+        'nhev',  # total number of hessian evaluations
+        'x_k',  # A tensor containing the last argument value found during the search. If the search converged, then
+        # this value is the argmin of the objective function.
+        'f_k',  # A tensor containing the value of the objective
+        # function at the `position`. If the search
+        # converged, then this is the (local) minimum of
+        # the objective function.
+        'g_k',  # A tensor containing the gradient of the
+        # objective function at the
+        # `final_position`. If the search converged
+        # the max-norm of this tensor should be
+        # below the tolerance.
+        'H_k'  # A tensor containing the inverse of the estimated Hessian.
+    ])
+
+
+def bfgs_minimize(func, x0, options=None):
+    """
+        Minimisation with BFGS. Inexact line search is performed satisfying both Wolfe conditions
+        (strong version of second one).
+        Args:
+            func: function to minimise, should take a single flat parameter, similar to numpy
+            x0: initial guess of parameter
+            options: A dictionary of solver options. All methods accept the following generic options:
+                maxiter : int or None which means inf
+                    Maximum number of iterations to perform. Depending on the
+                    method each iteration may use several function evaluations.
+                analytic_initial_hessian: bool,
+                    whether to use JAX to get hessian for the initialisation, otherwise it's eye.
+                g_tol: float
+                    stopping criteria is |grad(func)|_2 \lt g_tol
+                ls_maxiter: int or None which means inf
+                    Maximum number of line search iterations to perform
+
+        Returns: BFGSResults
+
+    """
+    if options is None:
+        options = dict()
+    maxiter: int = options.get('maxiter', None)
```

Comment:
Use ``Optional[int]`` to avoid possible pytype errors.

Reply:
Done.